packages/compass-import-export/src/export/gather-fields.ts

```diff
@@ -0,0 +1,169 @@
+  filter,
+  sampleSize,
+}: AnalyzeSchemaOptions) {
+  try {
+    // TODO(COMPASS-6426): Should we use the other aspects of the query in this sample. (project/sort/limit/skip)
+    const docs = await dataService.sample(
+      ns,
+      {
+        query: filter,
+        ...(sampleSize
+          ? {
+              size: sampleSize,
+            }
+          : {}),
+      },
+      {
+        promoteValues: false,
+      },
+      {
+        abortSignal,
+      }
+    );
+    const schemaData = await mongodbSchema(docs);
+    schemaData.fields = schemaData.fields.filter(
+      ({ path }: { path: string }) => !isInternalFieldPath(path)
+    );
+
+    return schemaData;
+  } catch (err) {
+    if (dataService.isCancelError(err)) {
+      debug('caught background operation terminated error', err);
+      return null;
+    }
+
+    debug('schema analysis failed', err);
+    throw err;
+  }
+}
+
+type GatherFieldsOptions = {
+  dataService: DataService;
+  ns: string;
+  abortSignal: AbortSignal;
+  filter?: Document;
+  progressCallback: (index: number) => void;
+  sampleSize: number;
+};
+
+// Array of path components. ie. { foo: { bar: { baz:  1 } } } results in ['foo', 'bar', 'baz']
+export type SchemaPath = string[];
+
+function schemaToPaths(
+  fields: SchemaField[],
+  parent: string[] = []
+): SchemaPath[] {
+  const paths: string[][] = [];
+
+  for (const field of fields) {
+    const path = [...parent, field.name];
+    paths.push(path);
+
+    // Recurse on doc.
+    const doc = field.types.find(
+      (f) =>
+        (
+          f as unknown as {
+            bsonType: string;
+          }
+        ).bsonType === 'Document'
+    );
+    if (doc) {
+      paths.push(
+        ...schemaToPaths(((doc as any).fields ?? []) as SchemaField[], path)
+      );
+    }
+
+    // Recurse on array.
+    const array = field.types.find(
+      (f) =>
+        (
+          f as unknown as {
+            bsonType: string;
+          }
+        ).bsonType === 'Array'
+    );
+    if (array) {
+      const arrayDoc = (array as any).types.find(
+        (f: { bsonType?: string }) => f.bsonType === 'Document'
+      );
+      if (arrayDoc) {
+        paths.push(...schemaToPaths(arrayDoc.fields ?? [], path));
+      }
+    }
+  }
+
+  return paths;
+}
+
+type Projection = {
+  [field: string]: boolean | Projection;
+};
+
+export function createProjectionFromSchemaFields(fields: SchemaPath[]) {
+  const projection: Projection = {};
+
+  for (const fieldPath of fields) {
+    let current = projection;
+    for (const [index, fieldName] of fieldPath.entries()) {
+      // Set the projection when it's the last index.
+      if (index === fieldPath.length) {
+        current[fieldName] = true;
+        break;
+      }
+
+      if (!current[fieldName]) {
+        current[fieldName] = {};
+      }
+
+      current = current[fieldName] as Projection;
+    }
+  }
+
+  return projection;
+}
+
+// TODO(COMPASS-6426): We will fill out the rest of this function and
```

Comment:
I'm actually now wondering if progress here would be overkill given that a) it is a sample and b) it is largely blocking: The only waits are around getting the sample and then running mongoSchema. mongodb-schema's parseSchema streams internally, but we're giving it an array of docs so it is Readable.from() with all the data in one go. And it has no progress mechanism yet anyway.

Reply:
Sounds good, if there's no progress to track let's remove it. We can do that in the COMPASS-6426 ticket work.